doc/developer/design/20210408_timelines.md

```diff
@@ -0,0 +1,96 @@
+Most sources use something around wall-clock now when generating that number, so asking about the state of various indexes at time X will incur at most a very short (< 1 second) wait for the data to be ready.
+
+A problem occurs when there is a source that generates timestamp numbers that may not be related to wall-clock now.
+The two in our system are Debezium consistency topics and CDCv2.
+Debezium consistency topics start their "timestamp" number at 1 and increment it for each transaction.
+If these sources are joined with wall-clock now sources, the dataflow layer will happily wait for the timestamps to be equal.
+For the Debezium sources, this would require waiting until there were as many transactions as milliseconds since the Unix epoch, which essentially will never happen.
+We should prevent this poor UX from happening to users.
+
+In addition to an infinite delay, there are cases where, even if the numbers were the same, they are meaningless when related to each other.
+For example, two separate Debezium consistency topics.
+Although they both start counting at 1 and thus have relatively close numbers, there's no meaning in the transaction counter, and joining data on that will not produce results that are meaningful.
+We should inform users when they attempt to join data from differing timelines that they have not asked us a useful question.
+
+The third concern is linearizability.
+We would like to provide that property when users issue SELECT queries.
+If we can distinguish the meaning of timestamps across data, we can provide linearizability across sources correctly.
+(This will be described with more detail in a later doc.)
+
+A solution to this problem is to inform the system what the timestamp number means for all data, and allow the system to reject queries that request data from multiple meanings.
+We introduce a thing called a timeline which is the meaning of the timestamp number.
+At the points in the system at which it is possible to join different sources together, we will return an error to the user if there is more than one timeline present.
+There are two places where this can occur: `CREATE VIEW` and `SELECT`.
+These have been taught to check that their expression depends at most one timeline.
+Other operations that can create dataflow indexes (TAIL, CREATE SOURCE, CREATE SINK, CREATE INDEX) operate on existing views, and are thus unable to create dataflows from multiple timelines.
+
+This is implemented by adding a `Timeline` enum, and requiring sources and tables to provide their `Timeline` to the Coordinator.
+When a VIEW or SELECT is planned, its MirRelationExpr is crawled to discover all timelines.
+
+There are currently three variants of `Timeline` (the enum docs describe why they are named as they are):
+
+- `RealTime` for anything producing timestamps near wall-clock now.
+- `Counter` for Debezium consistency topics.
+- `External` for CDCv2 sources.
```

Comment:
This may be coming / clearer in the near future, but it would be great if this were e.g.
```
enum Timeline {
    RealTime,
    Counter(String),
    External(String),
}
```
or something similar, so that it is clear (if it is the case) that we don't treat all `Counter` and `External` timelines as the same.

Reply:
Counter already comes with a broker+topic struct to tell them apart. External doesn't, under the assumption that users know what their timestamps mean and if they can be joined together.